dandi/download.py

```diff
@@ -706,15 +707,29 @@ def _download_file(
 
     # TODO: how do we discover the total size????
     # TODO: do not do it in-place, but rather into some "hidden" file
+    downloaded = 0
     for attempt in range(3):
         try:
-            downloaded = 0
+            if downloaded:
+                downloaded = max(downloaded - 2 * MAX_CHUNK_SIZE, 0)
             if digester:
                 downloaded_digest = digester()  # start empty
             warned = False
             # I wonder if we could make writing async with downloader
-            with open(path, "wb") as writer:
-                for block in downloader():
+            with open(path, "a+b") as writer:
+                first = True
+                for block in downloader(start_at=downloaded):
+                    if first:
+                        writer.seek(downloaded)
+                        if downloaded:
+                            blob1 = writer.read(len(block))
+                            if blob1 == block:
+                                downloaded += len(block)
```

Comment:
since succesful, and we went back two chunks, either here we would need to refeed a new digester with previously downloaded data, or may be just avoid updating digester below for redownloaded block?  without it -- it would end up with an incorrect digest I think, or am I wrong?

Reply:
Commit ba8767f should address this.

Reply:
thanks ... Going through the code I think that should work indeed!   But logic is a bit tricky (at least to me).  So I think we just need a good number of tests on some file longer than 3 chunks, and have 3 runs with download interruption on 1st chunk, then on 2nd, then on 3rd.  And then some test with fails on multiple chunks (so may be if file was 5 chunks, fail on 3 rd and 5 th).  Just to see that all aspects of restart work out nicely.

Reply:
Do you have any suggestions on how to simulate download interruptions?  The best thing I've been able to find for this is https://github.com/Shopify/toxiproxy, but it doesn't handle TLS certificates intelligently, and either way I've been unable to actually connect to its proxy in a minimal setup.

Reply:
Isn't our docker instance run over a straight http without any encryption? I would have just tried to mock shim the requests streaming "thing" to abort with some code reaching the desired chunk

Reply:
I'm not clear on exactly what error should be raised when simulating a broken connection.  `HTTPError` (the only thing the download code currently catches) is wrong, as that's only raised in response to 4xx and 5xx status codes, which occur at the beginning of a response.  [This SO post](https://stackoverflow.com/a/25862433/744178) says that a lost connection while iterating through `iter_content` results in a `socket.timeout` error, but I was unable to reproduce this in a minimal example in my own environment; when I tried, the script just hung for 15 minutes and then exited as though `iter_content` had ended naturally.

Reply:
I was thinking about 408 Request Timeout . Anyways - this is primarily to test out code/logic and ability to restart from a given point. I wouldn't necessarily aim to 100% match the real life scenario at this point

Reply:
What URL should be passed to `download()` to make it download a given Dandiset from the Docker Compose instance?  If I use `http://localhost:8079/dandiset/{dandi_id}/draft`, dandi-cli makes a HEAD request to the URL which messes with the mocked `iter_content()`.  If I use `http://localhost:8086/#/dandiset/{dandi_id}/draft`, I get `'NoneType' object has no attribute 'endswith'`.

Reply:
probably no need for entire dandiset download -- a single URL directly to the file should be enough.

"drafts" would still go through the girder, and only released ones (with a version) through new API.So they should be the "easiest" to try.  But we have not probably interfaced yet the API for "publish". you could find `/publish` end point at http://publish.dandiarchive.org/swagger/ (there is also #122 but I think @satra did not get to trigger it yet).  So might need need to be done first to test "our" (not girder client) direct downloader via requests.

You can mock exclusively that `/dandisets/{version__dandiset__pk}/versions/{version__version}/assets/{uuid}/download/` end point (eg. probably just pass through anything which `not re.search(query, '/versions/.*/assets/.*/download/)`.  You would just need to discover UUID explicitly (via a call which downloader user to get a list of assets) or indeed, as I guess you have tried, download an entire dandiset.

You could similarly pass through for whatever doesn't look like girder's request for a file (do not have a URL handy but also you could see how it looks like if you just print those which are sent out to requests by girder client).

Reply:
OK, I'm getting a bit lost.  Just how do you envision this test being structured?  Note that `responses` does not support raising an error in the middle of a streaming response, if that's what you were getting at.

Reply:
Hm... What then happens if eg connection is lost? If some other exception - let's raise that one?

Reply:
[As I said above](https://github.com/dandi/dandi-cli/pull/189#discussion_r473018715), when I tried inducing a lost connection with a test script (by turning off my WiFi), there was no error.  However, an SO post from six years ago says that such an event should raise a `socket.timeout`.

Even if we do determine the proper error to raise, there's still the matter of what to mock and how so that the error gets raised at the right point.

Reply:
hm... I have also failed so far (by disabling/enabling WiFi) to trigger requests to give up and raise some exception (but I didn't wait for too long I guess).  I have not realized that requests is really trying hard to maintain that stream.  But I doubt it would try going back and resume if connection does get interrupted somehow (e.g. jumping from one network to another?).

As for what to mock, I would have mocked `get_download_file_iter` which we have in both girder and dandiapi.  Mock shim them would obtain original generator function, but instead of returning that one, would return one which would be a generator which would yield from the original one until it hits the chunk it wants to raise on (implementation of `get_download_file_iter` in dandicli is easier to grasp - there is an explicit loop there).  I would for now raise `socket.timeout` SO talks about.

Related to all this is #198 where also "resume" would be used.

Reply:
I still can't figure out what URL to pass to `download()`.  The mock no longer interferes with `http://localhost:8079/dandiset/{dandi_id}/draft`, but now the request to that URL is returning 500, presumably because of the `GIRDER_URL` issue discovered in #186 that we didn't completely address because editing `serve.py` wasn't necessary at the time.